In [4]:
#x = 1
#x = 2.0
#x = "sddstgre34234"
x = True

type(x)

bool

## ciao

In [ ]:
## [] lista
## {:} dizionario
## () tupla
## --> numpy -->array ([])

In [5]:
pippo = [1,2,4854,849,'dgfd',308.2,True]
print(pippo)

[1, 2, 4854, 849, 'dgfd', 308.2, True]


In [9]:
lista = [3.0,6.0,9.0,12.0]
lista*2

[3.0, 6.0, 9.0, 12.0, 3.0, 6.0, 9.0, 12.0]

# Numpy

In [11]:
import numpy as np

In [12]:
country = np.array(['USA', 8 , 'UK', 'Celestopoli', 'India', 'China'])
print(country)

['USA' '8' 'UK' 'Celestopoli' 'India' 'China']


In [13]:
x = np.array([[2,3,1,0]])

In [14]:
x.shape

(1, 4)

In [17]:
y = np.array([[2,3,1,0],
              [1,1,1,1]])
y.shape

(2, 4)

In [18]:
np.zeros((5, 4))

array([[0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.]])

In [20]:
k =np.zeros((17, 2))
k

array([[0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.]])

In [21]:
k.shape

(17, 2)

In [ ]:
k

In [22]:
np.arange(10)

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [23]:
np.arange(2, 10, dtype=float) # default strp=1

array([2., 3., 4., 5., 6., 7., 8., 9.])

In [24]:
np.linspace(1., 4., 6)

array([1. , 1.6, 2.2, 2.8, 3.4, 4. ])

In [25]:
x = np.array([3, 6, 9, 12])

In [26]:
x/3.0

array([1., 2., 3., 4.])

In [30]:
print(x)

[ 3  6  9 12]


In [31]:
type(x)

numpy.ndarray

In [28]:
z = np.array([2,3,1,0])

In [29]:
np.sort(z)

array([0, 1, 2, 3])

In [30]:
a = np.array([[1, 2, 3], [4, 5, 6]])
b = np.array([[9, 8, 7], [6, 5, 4]])

In [31]:
c = np.concatenate((a, b))
c

array([[1, 2, 3],
       [4, 5, 6],
       [9, 8, 7],
       [6, 5, 4]])

In [ ]:
c.shape

In [ ]:
np.ones((2, 2), dtype=bool)

In [ ]:
z = np.array([[1, 0, 1], [0, 0, 0]],dtype=bool)
z

In [32]:
## Conversion to an array:

import numpy as np
my_list = [1, 2, 3, 4, 5, 6, 7, 8]
my_list = np.array(my_list)
#type(my_list)
my_list

array([1, 2, 3, 4, 5, 6, 7, 8])

In [35]:
my_tuple = ([8, 4, 6], [1, 2, 3])
type(my_tuple)

tuple

In [39]:
my_tuple = np.array(my_tuple)
my_tuple

array([[8, 4, 6],
       [1, 2, 3]])

#### Selection

In [37]:
my_tuple = ([8, 4, [6,0]], [1, 2, 3])
type(my_tuple)

tuple

In [44]:
my_tuple[0][2][1]

0

In [46]:
# dizionario??

In [45]:
a = {'key':3}
a

{'key': 3}

In [46]:
## vincolo per costruzione di dati strutturati
data = {'uno':[1,2,3,4,5,6],
        'due':[6,7,8,9,0,3]
        }

In [55]:
import pandas as pd

pd.DataFrame(data)

uno  due
0    1    6
1    2    7
2    3    8
3    4    9
4    5    0
5    6    3

# Pandas Dataframe

<img src="https://frenzy86.s3.eu-west-2.amazonaws.com/IFAO/dataframe.png" width="1200">

In [47]:
import numpy as np
import pandas as pd

In [49]:
np.random.seed(667)
mydata = np.random.randint(0,98,(4,3))

In [50]:
mydata

array([[85, 15, 89],
       [ 3, 90,  2],
       [64,  0, 36],
       [47, 17, 86]])

In [53]:
mydata.shape

(4, 3)

In [51]:
myindex = ['CA','NY','AZ','TX']

In [52]:
mycolumns = ['Jan','Feb','Mar']

In [63]:
df = pd.DataFrame(data=mydata)
df

0   1   2
0   85  15  89
1  100   3  90
2    2  64   0
3   36  99  47

In [54]:
df = pd.DataFrame(data=mydata,
                  index=myindex
                  )
df

0   1   2
CA  85  15  89
NY   3  90   2
AZ  64   0  36
TX  47  17  86

In [55]:
len(mycolumns)

3

In [56]:
df = pd.DataFrame(data=mydata,
                  index=myindex,
                  columns=mycolumns
                  )
df 

Jan  Feb  Mar
CA   85   15   89
NY    3   90    2
AZ   64    0   36
TX   47   17   86

In [57]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4 entries, CA to TX
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   Jan     4 non-null      int32
 1   Feb     4 non-null      int32
 2   Mar     4 non-null      int32
dtypes: int32(3)
memory usage: 80.0+ bytes


## Read external data tables

In [58]:
!wget https://frenzy86.s3.eu-west-2.amazonaws.com/IFAO/tips_complete.csv

"wget" non � riconosciuto come comando interno o esterno,
 un programma eseguibile o un file batch.


In [68]:
ls

sample_data/  tips_complete.csv


In [59]:
df = pd.read_csv('tips_complete.csv')
df

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2              8.49   
1         10.34  1.66    Male     No   Sun  Dinner     3              3.45   
2         21.01  3.50    Male     No   Sun  Dinner     3              7.00   
3         23.68  3.31    Male     No   Sun  Dinner     2             11.84   
4         24.59  3.61  Female     No   Sun  Dinner     4              6.15   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

             Payer Name         CC Number Payment ID  
0    Christy Cunningham  3560325168603410    Sun2959  
1        Douglas Tucker  4478071379779230    Sun4608  
2        Travis Walters  6011812112971322    Sun4458  
3      Nathaniel Harris  4676137647685994    Sun5260  
4          Tonya Carter  4832732618637221    Sun2251  
..                  ...               ...        ...  
239       Michael Avila  5296068606052842    Sat2657  
240      Monica Sanders  3506806155565404    Sat1766  
241          Keith Wong  6011891618747196    Sat3880  
242        Dennis Dixon     4375220550950      Sat17  
243     Michelle Hardin  3511451626698139    Thur672  

[244 rows x 11 columns]

----
About this DataSet (in case you are interested)

* Description
    * One waiter recorded information about each tip he received over a period of a few months working in one restaurant. He collected several variables:

* Format
    * A data frame with 244 rows and 7 variables

* Details
    * tip in dollars,
    * bill in dollars,
    * sex of the bill payer,
    * whether there were smokers in the party,
    * day of the week,
    * time of day,
    * size of the party.

In all he recorded 244 tips.
    
* Note: I created some additional columns with Fake data, including Name, CC Number, and Payment ID.

----

# DataFrames

In [60]:
df.columns

Index(['total_bill', 'tip', 'sex', 'smoker', 'day', 'time', 'size',
       'price_per_person', 'Payer Name', 'CC Number', 'Payment ID'],
      dtype='object')

In [61]:
df.index

RangeIndex(start=0, stop=244, step=1)

In [63]:
df.head(15)

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0        16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1        10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2        21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3        23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4        24.59  3.61  Female     No  Sun  Dinner     4              6.15   
5        25.29  4.71    Male     No  Sun  Dinner     4              6.32   
6         8.77  2.00    Male     No  Sun  Dinner     2              4.38   
7        26.88  3.12    Male     No  Sun  Dinner     4              6.72   
8        15.04  1.96    Male     No  Sun  Dinner     2              7.52   
9        14.78  3.23    Male     No  Sun  Dinner     2              7.39   
10       10.27  1.71    Male     No  Sun  Dinner     2              5.14   
11       35.26  5.00  Female     No  Sun  Dinner     4              8.82   
12       15.42  1.57    Male     No  Sun  Dinner     2              7.71   
13       18.43  3.00    Male     No  Sun  Dinner     4              4.61   
14       14.83  3.02  Female     No  Sun  Dinner     2              7.42   

            Payer Name         CC Number Payment ID  
0   Christy Cunningham  3560325168603410    Sun2959  
1       Douglas Tucker  4478071379779230    Sun4608  
2       Travis Walters  6011812112971322    Sun4458  
3     Nathaniel Harris  4676137647685994    Sun5260  
4         Tonya Carter  4832732618637221    Sun2251  
5           Erik Smith   213140353657882    Sun9679  
6   Kristopher Johnson  2223727524230344    Sun5985  
7          Robert Buck  3514785077705092    Sun8157  
8      Joseph Mcdonald  3522866365840377    Sun6820  
9        Jerome Abbott  3532124519049786    Sun3775  
10       William Riley      566287581219    Sun2546  
11        Diane Macias  4577817359320969    Sun6686  
12     Chad Harrington      577040572932    Sun1300  
13        Joshua Jones  6011163105616890    Sun2971  
14       Vanessa Jones    30016702287574    Sun3848

In [65]:
df.tail()

total_bill   tip     sex smoker   day    time  size  price_per_person  \
239       29.03  5.92    Male     No   Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

          Payer Name         CC Number Payment ID  
239    Michael Avila  5296068606052842    Sat2657  
240   Monica Sanders  3506806155565404    Sat1766  
241       Keith Wong  6011891618747196    Sat3880  
242     Dennis Dixon     4375220550950      Sat17  
243  Michelle Hardin  3511451626698139    Thur672

In [66]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   total_bill        244 non-null    float64
 1   tip               244 non-null    float64
 2   sex               244 non-null    object 
 3   smoker            244 non-null    object 
 4   day               244 non-null    object 
 5   time              244 non-null    object 
 6   size              244 non-null    int64  
 7   price_per_person  244 non-null    float64
 8   Payer Name        244 non-null    object 
 9   CC Number         244 non-null    int64  
 10  Payment ID        244 non-null    object 
dtypes: float64(3), int64(2), object(6)
memory usage: 21.1+ KB


In [67]:
len(df)

244

In [68]:
df.describe()

total_bill         tip        size  price_per_person     CC Number
count  244.000000  244.000000  244.000000        244.000000  2.440000e+02
mean    19.785943    2.998279    2.569672          7.888197  2.563496e+15
std      8.902412    1.383638    0.951100          2.914234  2.369340e+15
min      3.070000    1.000000    1.000000          2.880000  6.040679e+10
25%     13.347500    2.000000    2.000000          5.800000  3.040731e+13
50%     17.795000    2.900000    2.000000          7.255000  3.525318e+15
75%     24.127500    3.562500    3.000000          9.390000  4.553675e+15
max     50.810000   10.000000    6.000000         20.270000  6.596454e+15

In [69]:
#df.describe().transpose()
df.describe().T

count          mean           std           min  \
total_bill        244.0  1.978594e+01  8.902412e+00  3.070000e+00   
tip               244.0  2.998279e+00  1.383638e+00  1.000000e+00   
size              244.0  2.569672e+00  9.510998e-01  1.000000e+00   
price_per_person  244.0  7.888197e+00  2.914234e+00  2.880000e+00   
CC Number         244.0  2.563496e+15  2.369340e+15  6.040679e+10   

                           25%           50%           75%           max  
total_bill        1.334750e+01  1.779500e+01  2.412750e+01  5.081000e+01  
tip               2.000000e+00  2.900000e+00  3.562500e+00  1.000000e+01  
size              2.000000e+00  2.000000e+00  3.000000e+00  6.000000e+00  
price_per_person  5.800000e+00  7.255000e+00  9.390000e+00  2.027000e+01  
CC Number         3.040731e+13  3.525318e+15  4.553675e+15  6.596454e+15

## Selection and Indexing

Let's learn how to retrieve information from a DataFrame.

### COLUMNS

We will begin be learning how to extract information based on the columns

In [70]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

#### Grab a Single Column

In [73]:
df['total_bill']

0      16.99
1      10.34
2      21.01
3      23.68
4      24.59
       ...  
239    29.03
240    27.18
241    22.67
242    17.82
243    18.78
Name: total_bill, Length: 244, dtype: float64

In [74]:
type(df['total_bill'])

pandas.core.series.Series

#### Grab Multiple Columns

In [75]:
# Note how its a python list of column names! Thus the double brackets.
df[['total_bill','tip']]

total_bill   tip
0         16.99  1.01
1         10.34  1.66
2         21.01  3.50
3         23.68  3.31
4         24.59  3.61
..          ...   ...
239       29.03  5.92
240       27.18  2.00
241       22.67  2.00
242       17.82  1.75
243       18.78  3.00

[244 rows x 2 columns]

#### Create New Columns

In [76]:
df['tip_percentage'] = 100* df['tip'] / df['total_bill']

In [77]:
df

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2              8.49   
1         10.34  1.66    Male     No   Sun  Dinner     3              3.45   
2         21.01  3.50    Male     No   Sun  Dinner     3              7.00   
3         23.68  3.31    Male     No   Sun  Dinner     2             11.84   
4         24.59  3.61  Female     No   Sun  Dinner     4              6.15   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

             Payer Name         CC Number Payment ID  tip_percentage  
0    Christy Cunningham  3560325168603410    Sun2959        5.944673  
1        Douglas Tucker  4478071379779230    Sun4608       16.054159  
2        Travis Walters  6011812112971322    Sun4458       16.658734  
3      Nathaniel Harris  4676137647685994    Sun5260       13.978041  
4          Tonya Carter  4832732618637221    Sun2251       14.680765  
..                  ...               ...        ...             ...  
239       Michael Avila  5296068606052842    Sat2657       20.392697  
240      Monica Sanders  3506806155565404    Sat1766        7.358352  
241          Keith Wong  6011891618747196    Sat3880        8.822232  
242        Dennis Dixon     4375220550950      Sat17        9.820426  
243     Michelle Hardin  3511451626698139    Thur672       15.974441  

[244 rows x 12 columns]

In [78]:
df['price_per_person_'] = df['total_bill'] / df['size']

In [79]:
df.columns = ['total_bill', 'tip', 'sex', 'smoker', 'day', 'time', 'size',
       'price_per_person', 'Payer Name', 'CC Number', 'Payment ID',
       'tip_percentage', 'price_per_person_']

In [ ]:
help(np.round)

Help on function round_ in module numpy:

round_(a, decimals=0, out=None)
    Round an array to the given number of decimals.
    
    See Also
    --------
    around : equivalent function; see for details.



#### Adjust Existing Columns

In [80]:
df

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2              8.49   
1         10.34  1.66    Male     No   Sun  Dinner     3              3.45   
2         21.01  3.50    Male     No   Sun  Dinner     3              7.00   
3         23.68  3.31    Male     No   Sun  Dinner     2             11.84   
4         24.59  3.61  Female     No   Sun  Dinner     4              6.15   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

             Payer Name         CC Number Payment ID  tip_percentage  \
0    Christy Cunningham  3560325168603410    Sun2959        5.944673   
1        Douglas Tucker  4478071379779230    Sun4608       16.054159   
2        Travis Walters  6011812112971322    Sun4458       16.658734   
3      Nathaniel Harris  4676137647685994    Sun5260       13.978041   
4          Tonya Carter  4832732618637221    Sun2251       14.680765   
..                  ...               ...        ...             ...   
239       Michael Avila  5296068606052842    Sat2657       20.392697   
240      Monica Sanders  3506806155565404    Sat1766        7.358352   
241          Keith Wong  6011891618747196    Sat3880        8.822232   
242        Dennis Dixon     4375220550950      Sat17        9.820426   
243     Michelle Hardin  3511451626698139    Thur672       15.974441   

     price_per_person_  
0             8.495000  
1             3.446667  
2             7.003333  
3            11.840000  
4             6.147500  
..                 ...  
239           9.676667  
240          13.590000  
241          11.335000  
242           8.910000  
243           9.390000  

[244 rows x 13 columns]

In [84]:
# Because pandas is based on numpy, we get awesome capabilities with numpy's universal functions!
df['price_per_person_'] = np.round(df['price_per_person_'],1)

In [85]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  tip_percentage  \
0  Christy Cunningham  3560325168603410    Sun2959        5.944673   
1      Douglas Tucker  4478071379779230    Sun4608       16.054159   
2      Travis Walters  6011812112971322    Sun4458       16.658734   
3    Nathaniel Harris  4676137647685994    Sun5260       13.978041   
4        Tonya Carter  4832732618637221    Sun2251       14.680765   

   price_per_person_  
0                8.5  
1                3.4  
2                7.0  
3               11.8  
4                6.2

#### Remove Columns

In [ ]:
# df.drop('tip_percentage',axis=1)

In [87]:
df = df.drop("price_per_person_",axis=1)

In [88]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  tip_percentage  
0  Christy Cunningham  3560325168603410    Sun2959        5.944673  
1      Douglas Tucker  4478071379779230    Sun4608       16.054159  
2      Travis Walters  6011812112971322    Sun4458       16.658734  
3    Nathaniel Harris  4676137647685994    Sun5260       13.978041  
4        Tonya Carter  4832732618637221    Sun2251       14.680765

# Index Basics

Before going over the same retrieval tasks for rows, let's build some basic understanding of the pandas DataFrame Index.

In [89]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  tip_percentage  
0  Christy Cunningham  3560325168603410    Sun2959        5.944673  
1      Douglas Tucker  4478071379779230    Sun4608       16.054159  
2      Travis Walters  6011812112971322    Sun4458       16.658734  
3    Nathaniel Harris  4676137647685994    Sun5260       13.978041  
4        Tonya Carter  4832732618637221    Sun2251       14.680765

In [90]:
df.index

RangeIndex(start=0, stop=244, step=1)

In [93]:
df.set_index('Payment ID')

total_bill   tip     sex smoker   day    time  size  \
Payment ID                                                        
Sun2959          16.99  1.01  Female     No   Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No   Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No   Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No   Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No   Sun  Dinner     4   
...                ...   ...     ...    ...   ...     ...   ...   
Sat2657          29.03  5.92    Male     No   Sat  Dinner     3   
Sat1766          27.18  2.00  Female    Yes   Sat  Dinner     2   
Sat3880          22.67  2.00    Male    Yes   Sat  Dinner     2   
Sat17            17.82  1.75    Male     No   Sat  Dinner     2   
Thur672          18.78  3.00  Female     No  Thur  Dinner     2   

            price_per_person          Payer Name         CC Number  \
Payment ID                                                           
Sun2959                 8.49  Christy Cunningham  3560325168603410   
Sun4608                 3.45      Douglas Tucker  4478071379779230   
Sun4458                 7.00      Travis Walters  6011812112971322   
Sun5260                11.84    Nathaniel Harris  4676137647685994   
Sun2251                 6.15        Tonya Carter  4832732618637221   
...                      ...                 ...               ...   
Sat2657                 9.68       Michael Avila  5296068606052842   
Sat1766                13.59      Monica Sanders  3506806155565404   
Sat3880                11.34          Keith Wong  6011891618747196   
Sat17                   8.91        Dennis Dixon     4375220550950   
Thur672                 9.39     Michelle Hardin  3511451626698139   

            tip_percentage  
Payment ID                  
Sun2959           5.944673  
Sun4608          16.054159  
Sun4458          16.658734  
Sun5260          13.978041  
Sun2251          14.680765  
...                    ...  
Sat2657          20.392697  
Sat1766           7.358352  
Sat3880           8.822232  
Sat17             9.820426  
Thur672          15.974441  

[244 rows x 11 columns]

In [94]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  tip_percentage  
0  Christy Cunningham  3560325168603410    Sun2959        5.944673  
1      Douglas Tucker  4478071379779230    Sun4608       16.054159  
2      Travis Walters  6011812112971322    Sun4458       16.658734  
3    Nathaniel Harris  4676137647685994    Sun5260       13.978041  
4        Tonya Carter  4832732618637221    Sun2251       14.680765

In [95]:
df = df.set_index('Payment ID')

In [96]:
df.head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   

            price_per_person          Payer Name         CC Number  \
Payment ID                                                           
Sun2959                 8.49  Christy Cunningham  3560325168603410   
Sun4608                 3.45      Douglas Tucker  4478071379779230   
Sun4458                 7.00      Travis Walters  6011812112971322   
Sun5260                11.84    Nathaniel Harris  4676137647685994   
Sun2251                 6.15        Tonya Carter  4832732618637221   

            tip_percentage  
Payment ID                  
Sun2959           5.944673  
Sun4608          16.054159  
Sun4458          16.658734  
Sun5260          13.978041  
Sun2251          14.680765

In [97]:
df = df.reset_index()

In [98]:
df.head()

Payment ID  total_bill   tip     sex smoker  day    time  size  \
0    Sun2959       16.99  1.01  Female     No  Sun  Dinner     2   
1    Sun4608       10.34  1.66    Male     No  Sun  Dinner     3   
2    Sun4458       21.01  3.50    Male     No  Sun  Dinner     3   
3    Sun5260       23.68  3.31    Male     No  Sun  Dinner     2   
4    Sun2251       24.59  3.61  Female     No  Sun  Dinner     4   

   price_per_person          Payer Name         CC Number  tip_percentage  
0              8.49  Christy Cunningham  3560325168603410        5.944673  
1              3.45      Douglas Tucker  4478071379779230       16.054159  
2              7.00      Travis Walters  6011812112971322       16.658734  
3             11.84    Nathaniel Harris  4676137647685994       13.978041  
4              6.15        Tonya Carter  4832732618637221       14.680765

### ROWS

Let's now explore these same concepts but with Rows.

In [99]:
df.head()

Payment ID  total_bill   tip     sex smoker  day    time  size  \
0    Sun2959       16.99  1.01  Female     No  Sun  Dinner     2   
1    Sun4608       10.34  1.66    Male     No  Sun  Dinner     3   
2    Sun4458       21.01  3.50    Male     No  Sun  Dinner     3   
3    Sun5260       23.68  3.31    Male     No  Sun  Dinner     2   
4    Sun2251       24.59  3.61  Female     No  Sun  Dinner     4   

   price_per_person          Payer Name         CC Number  tip_percentage  
0              8.49  Christy Cunningham  3560325168603410        5.944673  
1              3.45      Douglas Tucker  4478071379779230       16.054159  
2              7.00      Travis Walters  6011812112971322       16.658734  
3             11.84    Nathaniel Harris  4676137647685994       13.978041  
4              6.15        Tonya Carter  4832732618637221       14.680765

In [ ]:
df = df.set_index('Payment ID')
df

total_bill   tip     sex smoker   day    time  size  \
Payment ID                                                        
Sun2959          16.99  1.01  Female     No   Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No   Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No   Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No   Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No   Sun  Dinner     4   
...                ...   ...     ...    ...   ...     ...   ...   
Sat2657          29.03  5.92    Male     No   Sat  Dinner     3   
Sat1766          27.18  2.00  Female    Yes   Sat  Dinner     2   
Sat3880          22.67  2.00    Male    Yes   Sat  Dinner     2   
Sat17            17.82  1.75    Male     No   Sat  Dinner     2   
Thur672          18.78  3.00  Female     No  Thur  Dinner     2   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun4608                 3.45      Douglas Tucker  4478071379779230  
Sun4458                 7.00      Travis Walters  6011812112971322  
Sun5260                11.84    Nathaniel Harris  4676137647685994  
Sun2251                 6.15        Tonya Carter  4832732618637221  
...                      ...                 ...               ...  
Sat2657                 9.68       Michael Avila  5296068606052842  
Sat1766                13.59      Monica Sanders  3506806155565404  
Sat3880                11.34          Keith Wong  6011891618747196  
Sat17                   8.91        Dennis Dixon     4375220550950  
Thur672                 9.39     Michelle Hardin  3511451626698139  

[244 rows x 10 columns]

#### Grab a Single Row

In [105]:
# Integer Based
df.iloc[:10][['total_bill','tip']]

total_bill   tip
0       16.99  1.01
1       10.34  1.66
2       21.01  3.50
3       23.68  3.31
4       24.59  3.61
5       25.29  4.71
6        8.77  2.00
7       26.88  3.12
8       15.04  1.96
9       14.78  3.23

In [108]:
df = df.set_index('Payment ID')

In [110]:
# Name Based
df.loc['Sun2959']

total_bill                       16.99
tip                               1.01
sex                             Female
smoker                              No
day                                Sun
time                            Dinner
size                                 2
price_per_person                  8.49
Payer Name          Christy Cunningham
CC Number             3560325168603410
tip_percentage                5.944673
Name: Sun2959, dtype: object

#### Grab Multiple Rows

In [111]:
df.iloc[0:4]

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   

            price_per_person          Payer Name         CC Number  \
Payment ID                                                           
Sun2959                 8.49  Christy Cunningham  3560325168603410   
Sun4608                 3.45      Douglas Tucker  4478071379779230   
Sun4458                 7.00      Travis Walters  6011812112971322   
Sun5260                11.84    Nathaniel Harris  4676137647685994   

            tip_percentage  
Payment ID                  
Sun2959           5.944673  
Sun4608          16.054159  
Sun4458          16.658734  
Sun5260          13.978041

In [112]:
df.loc[['Sun2959','Sun5260']]

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   

            price_per_person          Payer Name         CC Number  \
Payment ID                                                           
Sun2959                 8.49  Christy Cunningham  3560325168603410   
Sun5260                11.84    Nathaniel Harris  4676137647685994   

            tip_percentage  
Payment ID                  
Sun2959           5.944673  
Sun5260          13.978041

#### Remove Row

Typically are datasets will be large enough that we won't remove rows like this since we won't know thier row location for some specific condition, instead, we drop rows based on conditions such as missing data or column values. The next lecture will cover this in a lot more detail.

In [113]:
df.head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   

            price_per_person          Payer Name         CC Number  \
Payment ID                                                           
Sun2959                 8.49  Christy Cunningham  3560325168603410   
Sun4608                 3.45      Douglas Tucker  4478071379779230   
Sun4458                 7.00      Travis Walters  6011812112971322   
Sun5260                11.84    Nathaniel Harris  4676137647685994   
Sun2251                 6.15        Tonya Carter  4832732618637221   

            tip_percentage  
Payment ID                  
Sun2959           5.944673  
Sun4608          16.054159  
Sun4458          16.658734  
Sun5260          13.978041  
Sun2251          14.680765

In [114]:
df.drop('Sun2959',axis=0).head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   
Sun9679          25.29  4.71    Male     No  Sun  Dinner     4   

            price_per_person        Payer Name         CC Number  \
Payment ID                                                         
Sun4608                 3.45    Douglas Tucker  4478071379779230   
Sun4458                 7.00    Travis Walters  6011812112971322   
Sun5260                11.84  Nathaniel Harris  4676137647685994   
Sun2251                 6.15      Tonya Carter  4832732618637221   
Sun9679                 6.32        Erik Smith   213140353657882   

            tip_percentage  
Payment ID                  
Sun4608          16.054159  
Sun4458          16.658734  
Sun5260          13.978041  
Sun2251          14.680765  
Sun9679          18.623962

In [ ]:
# Error if you have a named index!
# df.drop(0,axis=0).head()

#### Insert a New Row

Pretty rare to add a single row like this. Usually you use pd.concat() to add many rows at once. You could use the .append() method with a list of pd.Series() objects, but you won't see us do this with realistic real-world data.

In [115]:
one_row = df.iloc[0]

In [116]:
one_row

total_bill                       16.99
tip                               1.01
sex                             Female
smoker                              No
day                                Sun
time                            Dinner
size                                 2
price_per_person                  8.49
Payer Name          Christy Cunningham
CC Number             3560325168603410
tip_percentage                5.944673
Name: Sun2959, dtype: object

In [117]:
type(one_row)

pandas.core.series.Series

In [118]:
df.tail()

total_bill   tip     sex smoker   day    time  size  \
Payment ID                                                        
Sat2657          29.03  5.92    Male     No   Sat  Dinner     3   
Sat1766          27.18  2.00  Female    Yes   Sat  Dinner     2   
Sat3880          22.67  2.00    Male    Yes   Sat  Dinner     2   
Sat17            17.82  1.75    Male     No   Sat  Dinner     2   
Thur672          18.78  3.00  Female     No  Thur  Dinner     2   

            price_per_person       Payer Name         CC Number  \
Payment ID                                                        
Sat2657                 9.68    Michael Avila  5296068606052842   
Sat1766                13.59   Monica Sanders  3506806155565404   
Sat3880                11.34       Keith Wong  6011891618747196   
Sat17                   8.91     Dennis Dixon     4375220550950   
Thur672                 9.39  Michelle Hardin  3511451626698139   

            tip_percentage  
Payment ID                  
Sat2657          20.392697  
Sat1766           7.358352  
Sat3880           8.822232  
Sat17             9.820426  
Thur672          15.974441

In [121]:
df.append(one_row).tail()

C:\Users\danie\AppData\Local\Temp\ipykernel_19724\112637673.py:1: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df.append(one_row).tail()


total_bill   tip     sex smoker   day    time  size  \
Payment ID                                                        
Sat1766          27.18  2.00  Female    Yes   Sat  Dinner     2   
Sat3880          22.67  2.00    Male    Yes   Sat  Dinner     2   
Sat17            17.82  1.75    Male     No   Sat  Dinner     2   
Thur672          18.78  3.00  Female     No  Thur  Dinner     2   
Sun2959          16.99  1.01  Female     No   Sun  Dinner     2   

            price_per_person          Payer Name         CC Number  \
Payment ID                                                           
Sat1766                13.59      Monica Sanders  3506806155565404   
Sat3880                11.34          Keith Wong  6011891618747196   
Sat17                   8.91        Dennis Dixon     4375220550950   
Thur672                 9.39     Michelle Hardin  3511451626698139   
Sun2959                 8.49  Christy Cunningham  3560325168603410   

            tip_percentage  
Payment ID                  
Sat1766           7.358352  
Sat3880           8.822232  
Sat17             9.820426  
Thur672          15.974441  
Sun2959           5.944673

--------

#### Extract an array

In [123]:
array_row = df.iloc[0:3].values
array_row

array([[16.99, 1.01, 'Female', 'No', 'Sun', 'Dinner', 2, 8.49,
        'Christy Cunningham', 3560325168603410, 5.9446733372572105],
       [10.34, 1.66, 'Male', 'No', 'Sun', 'Dinner', 3, 3.45,
        'Douglas Tucker', 4478071379779230, 16.054158607350097],
       [21.01, 3.5, 'Male', 'No', 'Sun', 'Dinner', 3, 7.0,
        'Travis Walters', 6011812112971322, 16.658733936220845]],
      dtype=object)

In [124]:
array_row.shape

(3, 11)